onnxruntime/core/optimizer/graph_transformer_utils.cc

```diff
@@ -41,14 +45,30 @@ std::vector<std::unique_ptr<RewriteRule>> GenerateRewriteRules(const Transformer
   return rules;
 }
 
-std::vector<TransformerProviderSet> GenerateTransformers(const TransformerLevel& level, 
+std::unique_ptr<RuleBasedGraphTransformer> GenerateRuleBasedGraphTransformer(const TransformerLevel& level,
+                                                                             const std::vector<std::string>* rules_to_enable,
+                                                                             std::string t_name) {
+  auto rewrite_rules_to_register = transformer_utils::GenerateRewriteRules(level, rules_to_enable);
+  if (rewrite_rules_to_register.empty()) {
+    return std::unique_ptr<RuleBasedGraphTransformer>{};
+  }
+
+  std::unique_ptr<RuleBasedGraphTransformer> rule_transformer =
+      std::make_unique<TopDownRuleBasedTransformer>(t_name + "_RuleBasedTransformer",
```

Comment:
I think this is our biggest debate :D 
Okay, maybe I will remove it :) For now :)

Reply:
Apply is provided by the base class (so that Graph::Resolve can be automatically called the minimum number of times). ApplyImpl is what the derived class needs.

---
In reply to: [267858400](https://github.com/Microsoft/onnxruntime/pull/670#discussion_r267858400) [](ancestors = 267858400)

Reply:
That's right. And frankly I don't find this confusing at all. But at every PR I open I get a comment for the TopDownRuleBasedTransformer, and I don't think it is worth arguing over. So let's keep it out until we have a need for a bottom up rule-based transformer.